src/zm_controller.cpp

```diff
@@ -6465,16 +6465,10 @@ void zmController::zclReadAttributesResponse(NodeInfo *node, const deCONZ::ApsDa
 
         if (cluster)
         {
-            for (uint i = 0; i < cluster->attributes().size(); i++)
+            for (auto &a : cluster->attributes())
             {
-                auto &a = cluster->attributes()[i];
-                if (a.id() == id)
+                if (a.id() == id && a.manufacturerCode() == zclFrame.manufacturerCode())

```

Comment:
Tricky, I think the former test of `a.isManufacturerSpecific() && a.manufacturerCode() != zclFrame.manufacturerCode()` is important, since `isManufactuerSpecific + manufacturerCode` needs to align.

However the old test was incomplete and misses the check the other way around.
To be strict the code needs two checks:

```C++
bool bIsManufacturerSpecific = (zclFrame.frameControl() & deCONZ::ZclFCManufacturerSpecific) != 0;
if (a.isManufacturerSpecific() != bIsManufacturerSpecific)
{
    continue;
}

if (a.isManufacturerSpecific() && a.manufacturerCode() != zclFrame.manufacturerCode())
{
    continue;
}
```

~~Note while leaving out the `isManufacturerSpecific()` test appears to work likely in most cases it relies on mfcode to be 0x0000, and if some creative vendor decides on using this as manufacturer specific value we can only handle this with above two checks and the frame control bit.~~

Hmm looking deeper the `ZclAttribute::isManufacturerSpecific()` right now is also just a test for `mfcode != 0`, were in contrast the `zclFrame::frameControl()` actually may carry a 0x0000 manufacturer specific thing. ... tricky  :thinking:

Reply:
> Hmm looking deeper the ZclAttribute::isManufacturerSpecific() right now is also just a test for mfcode != 0, were in contrast the zclFrame::frameControl() actually may carry a 0x0000 manufacturer specific thing. ... tricky 🤔

Saw that as well and that is probably the main reason why I left it out. I get your point here, but wonder if and to what extend that is a realistic scenario to occur. Also the attribute side has no means to express it is manufacturer specitic other than an mfc !=0 🙂

Reply:
Indeed I'm also not aware of any device doing this from memory, Tuya made me a bit paranoid over time :D
So I think we can go with your current code here perhaps with a small comment to check later if anything weird shows up.

Reply:
You got it 👍